packages/blocks/image/src/components/Media.tsx

```diff
@@ -0,0 +1,366 @@
+import {
+  BlockProtocolEntity,
+  BlockProtocolLinkGroup,
+  BlockProtocolUpdateEntitiesAction,
+  BlockProtocolUploadFileFunction,
+} from "blockprotocol";
+import { BlockComponent } from "blockprotocol/react";
+import React, {
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from "react";
+import { unstable_batchedUpdates } from "react-dom";
+import { ErrorAlert } from "./ErrorAlert";
+import { MediaWithCaption } from "./MediaWithCaption";
+import { UploadMediaForm } from "./UploadMediaForm";
+
+// https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#recursive-conditional-types
+type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;
+type FileType = Awaited<ReturnType<BlockProtocolUploadFileFunction>>;
+type UpdateEntitiesMediaAction = {
+  initialCaption?: string;
+  initialWidth?: number;
+  file?: FileType;
+};
+
+const useDefaultState = <T extends any>(
```

Comment:
To capture the fact it won't work for primitive values (could have a comment instead)
```suggestion
const useDefaultState = <
  T extends number | string | boolean | null | undefined,
>(
```

Reply:
Good idea – done!